rlclientlib/deserialization/joined_log_parser.cc

```diff
@@ -0,0 +1,92 @@
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include "err_constants.h"
+#include "example_joiner.h"
+#include "joined_log_parser.h"
+
+namespace v2 = reinforcement_learning::messages::flatbuff::v2;
+namespace err = reinforcement_learning::error_code;
+
+JoinedLogParser::JoinedLogParser(const std::string &initial_command_line)
+    : example_joiner(VW::make_unique<ExampleJoiner>(initial_command_line)) {}
+
+JoinedLogParser::~JoinedLogParser() = default;
+
+// TODO make better error messages
+// TODO check endianness
+int JoinedLogParser::read_and_deserialize_file(const std::string &file_name) {
+  std::ifstream fs(file_name.c_str(), std::ifstream::binary);
+  std::vector<char> buffer(4, 0);
+  const std::vector<char> magic = {'V', 'W', 'F', 'B'};
+
+  // read the 4 magic bytes
+  fs.read(buffer.data(), buffer.size());
+  if (buffer != magic) {
+    return err::file_read_error;
+  }
+
+  // read the version
+  fs.read(buffer.data(), buffer.size());
+  if (static_cast<int>(buffer[0]) != 1) {
+    return err::file_read_error;
+  }
+
+  // payload type, check for header
+  unsigned int payload_type;
+  fs.read((char *)(&payload_type), sizeof(payload_type));
+  if (payload_type != MSG_TYPE_HEADER) {
+    return err::file_read_error;
```

Comment:
We'll have to harden this later using ApiStatus and include better error messages.